backend/src/db/migrations/20200312140328-bulk_upload_to_s3.js

```diff
@@ -0,0 +1,103 @@
+import { getDriver } from '../../db/neo4j'
+import { existsSync, createReadStream } from 'fs'
+import path from 'path'
+import { S3 } from 'aws-sdk'
+import mime from 'mime-types'
+import { s3Configs } from '../../config'
+
+export const description = `
+Upload all image files to a S3 compatible object storage in order to reduce
+load on our backend.
+`
+
+export async function up(next) {
+  const driver = getDriver()
+  const session = driver.session()
+  const transaction = session.beginTransaction()
+
+  const {
+    AWS_ENDPOINT: endpoint,
+    AWS_REGION: region,
+    AWS_BUCKET: Bucket,
+    S3_CONFIGURED,
+  } = s3Configs
+
+  if (!S3_CONFIGURED) {
+    // eslint-disable-next-line no-console
+    console.log('No S3 given, cannot upload image files')
+    return
+  }
+
+  const s3 = new S3({ region, endpoint })
+  try {
+    // Implement your migration here.
+    const { records } = await transaction.run('MATCH (image:Image) RETURN image.url as url')
+    let urls = records.map((r) => r.get('url'))
+    urls = urls.filter((url) => url.startsWith('/uploads'))
+    const locations = await Promise.all(
+      urls
+        .map((url) => {
+          return async () => {
+            const { pathname } = new URL(url, 'http://example.org')
+            const fileLocation = path.join(__dirname, `../../../public/${pathname}`)
+            const s3Location = `original${pathname}`
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-03-26T14:47:35Z" title="Thursday, March 26th 2020, 3:47:35 pm +01:00">Mar 26, 2020</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -32,39 +32,38 @@ export async function up(next) {
   try {
     // Implement your migration here.
     const { records } = await transaction.run('MATCH (image:Image) RETURN image.url as url')
-    let urls = records.map(r => r.get('url'))
-    urls = urls.filter(url => url.startsWith('/uploads'))
+    let urls = records.map((r) => r.get('url'))
+    urls = urls.filter((url) => url.startsWith('/uploads'))
     const locations = await Promise.all(
       urls
-        .map(url => {
+        .map((url) => {
           return async () => {
             const { pathname } = new URL(url, 'http://example.org')
             const fileLocation = path.join(__dirname, `../../../public/${pathname}`)
             const s3Location = `original${pathname}`
```


I see that the images now live on our community staging. Do we want to have them in a folder `uploads/...`? I would say, we should omit that folder name, no?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2020-03-26T14:49:18Z" title="Thursday, March 26th 2020, 3:49:18 pm +01:00">Mar 26, 2020</time>_

---


Or, maybe not. We can clearly distinguish the images that have been uploaded with this migration. I case we have to do some disaster recovery, this could come in handy.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-03-26T15:13:16Z" title="Thursday, March 26th 2020, 4:13:16 pm +01:00">Mar 26, 2020</time>_

---


the reason it lives there is because it's part of the `Image.url`
yeah, agree with the second part of your message, let's KISS